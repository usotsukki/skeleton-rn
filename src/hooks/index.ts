export { default as useLoadFonts } from './useLoadFonts'
export { default as useConnectionStatus } from './useConnectionStatus'
