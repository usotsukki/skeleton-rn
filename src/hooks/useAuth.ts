import auth, { FirebaseAuthTypes } from '@react-native-firebase/auth'
import { useMutation } from '@tanstack/react-query'
import { useEffect } from 'react'
import { create } from 'zustand'
import { createJSONStorage, persist, PersistOptions, StateStorage } from 'zustand/middleware'
import {
	createUser as authCreateUser,
	signIn as authSignIn,
	signOut as authSignOut,
	getAppleAuthCredential,
	getGoogleAuthCredential,
	handleAuthErrors,
	signInWithProvider,
} from '@app/api/auth'
import { authStorage } from '@app/storage'

interface AuthState {
	user: FirebaseAuthTypes.User | null
}

interface AuthStore extends AuthState {
	setUser: (user: FirebaseAuthTypes.User | null) => void
}

const persistStorage: StateStorage = {
	setItem: (name, value) => authStorage.set(name, value),
	getItem: name => authStorage.getString(name) || null,
	removeItem: name => authStorage.delete(name),
}

const persistConfig: PersistOptions<AuthStore> = {
	name: 'auth',
	storage: createJSONStorage(() => persistStorage),
}

export const useAuthStore = create(
	persist(
		set => ({
			user: null,
			setUser: (user: FirebaseAuthTypes.User | null) => set({ user }),
		}),
		persistConfig,
	),
)

export const useAuthListener = () => {
	const setUser = useAuthStore(state => state.setUser)

	useEffect(() => {
		const unsubscribe = auth().onAuthStateChanged(setUser)
		return unsubscribe
	}, [])
}

const useAuth = () => {
	const { mutate: createUser, isPending: isCreateUserPending } = useMutation({
		mutationFn: ({ email, password }: { email: string; password: string }) => {
			return authCreateUser(email, password)
		},
		onError: handleAuthErrors,
	})

	const { mutate: signIn, isPending: isSignInPending } = useMutation({
		mutationFn: ({ email, password }: { email: string; password: string }) => {
			return authSignIn(email, password)
		},
		onError: handleAuthErrors,
	})

	const { mutate: signInWithGoogle, isPending: isSignInWithGooglePending } = useMutation({
		mutationFn: () => signInWithProvider(getGoogleAuthCredential),
		onError: handleAuthErrors,
	})

	const { mutate: signInWithApple, isPending: isSignInWithApplePending } = useMutation({
		mutationFn: () => signInWithProvider(getAppleAuthCredential),
		onError: handleAuthErrors,
	})

	const { mutate: signOut, isPending: isSignOutPending } = useMutation({
		mutationFn: () => authSignOut(),
		onError: handleAuthErrors,
	})

	const loading =
		isSignInPending || isSignInWithGooglePending || isSignInWithApplePending || isCreateUserPending || isSignOutPending

	return { signIn, signInWithGoogle, signInWithApple, createUser, signOut, loading }
}

export default useAuth
