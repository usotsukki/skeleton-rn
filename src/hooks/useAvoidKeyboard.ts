import { useFocusEffect } from 'expo-router'
import { useCallback } from 'react'
import { AvoidSoftInput } from 'react-native-avoid-softinput'

/**
 * Has to be used with scrollable components
 */
const useAvoidKeyboard = () => {
	const onFocusEffect = useCallback(() => {
		AvoidSoftInput.setShouldMimicIOSBehavior(true)
		AvoidSoftInput.setEnabled(true)
		return () => {
			AvoidSoftInput.setEnabled(false)
			AvoidSoftInput.setShouldMimicIOSBehavior(false)
		}
	}, [])

	useFocusEffect(onFocusEffect)
}

export default useAvoidKeyboard
