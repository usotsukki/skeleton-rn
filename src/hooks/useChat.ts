import { useMutation } from '@tanstack/react-query'
import { fetchChatCompletion } from '@app/api/chat'

const useChats = () => {
	return useMutation({
		mutationFn: async (userMessage: string) =>
			fetchChatCompletion({
				message: userMessage,
			}),
	})
}

export default useChats
