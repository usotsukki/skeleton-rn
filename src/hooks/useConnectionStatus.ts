import * as NetInfo from '@react-native-community/netinfo'
import { isNil } from 'lodash'
import { useCallback, useEffect, useState } from 'react'
import { logger } from '@app/utils'

const useConnectionStatus = () => {
	const [isConnected, setIsConnected] = useState(true)

	const onChange: NetInfo.NetInfoChangeHandler = useCallback(
		({ isInternetReachable }) => {
			if (!isNil(isInternetReachable)) {
				setIsConnected(isInternetReachable)
				if (!isConnected && isInternetReachable) {
					logger.info('Internet connection restored')
				} else if (!isInternetReachable) {
					logger.info('Internet connection lost, entering offline mode')
				}
			}
		},
		[isConnected],
	)

	useEffect(() => {
		const unsubscribe = NetInfo.addEventListener(onChange)
		return unsubscribe
	}, [onChange])

	return isConnected
}

export default useConnectionStatus
