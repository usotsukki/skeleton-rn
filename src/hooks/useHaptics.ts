import * as Haptics from 'expo-haptics'
import { useCallback } from 'react'

const useHaptics = () => {
	const triggerImpact = useCallback((style: Haptics.ImpactFeedbackStyle = Haptics.ImpactFeedbackStyle.Medium) => {
		Haptics.impactAsync(style)
	}, [])

	const triggerNotification = useCallback(
		(type: Haptics.NotificationFeedbackType = Haptics.NotificationFeedbackType.Success) => {
			Haptics.notificationAsync(type)
		},
		[],
	)

	const triggerSelection = useCallback(() => {
		Haptics.selectionAsync()
	}, [])

	return {
		triggerImpact,
		triggerNotification,
		triggerSelection,
	}
}

export default useHaptics
