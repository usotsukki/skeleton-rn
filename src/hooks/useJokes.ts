import { useQuery } from '@tanstack/react-query'
import { fetchRandomJoke } from '@app/api/jokes'

const useJokes = () => {
	return useQuery({
		queryKey: ['joke'],
		queryFn: fetchRandomJoke,
		refetchOnReconnect: 'always',
		experimental_prefetchInRender: true,
	})
}

export default useJokes
