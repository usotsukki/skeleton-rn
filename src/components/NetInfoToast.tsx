import React, { useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet } from 'react-native'
import { Incubator } from 'react-native-ui-lib'
import { useConnectionStatus } from '@app/hooks'

const NetInfoToast = () => {
	const { t } = useTranslation('translation', { keyPrefix: 'info' })
	const [visible, setVisible] = useState(false)

	const internetConnection = useConnectionStatus()

	useEffect(() => {
		if (!internetConnection) {
			setVisible(true)
		}
		if (internetConnection) {
			setVisible(false)
		}
	}, [internetConnection])

	return (
		<Incubator.Toast
			testID="toast-netinfo"
			zIndex={5}
			backgroundColor="rgba(0,0,0,0.9)"
			message={t('offline')}
			centerMessage
			messageStyle={styles.message}
			icon={require('@assets/png/offline.png')}
			visible={visible}
			swipeable
			position={'top'}
		/>
	)
}

const styles = StyleSheet.create({
	message: { color: 'white' },
})

export default NetInfoToast
