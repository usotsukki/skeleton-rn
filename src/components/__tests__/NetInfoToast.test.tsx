import { act, fireEvent, render, screen } from '@testing-library/react-native'
import React from 'react'
import NetInfoToast from '@app/components/NetInfoToast'
import * as hooks from '@app/hooks'

const mockedTranslate = jest.fn().mockReturnValue("You're currently offline")

const mockConnectionStatus = (status: boolean) => {
	jest.spyOn(hooks, 'useConnectionStatus').mockImplementation(() => status)
}

jest.mock('react-i18next', () => ({
	useTranslation: () => ({
		t: mockedTranslate,
	}),
}))

jest.mock('@app/hooks', () => ({
	useConnectionStatus: jest.fn(),
}))

describe('NetInfoToast', () => {
	afterEach(() => {
		jest.clearAllMocks()
	})

	it('should display the toast if the user is offline', () => {
		jest.useFakeTimers()
		mockConnectionStatus(false)
		render(<NetInfoToast />)

		act(() => jest.runAllTimers())

		expect(screen.getByTestId('toast-netinfo')).toBeTruthy()
		expect(mockedTranslate).toHaveBeenCalledWith('offline')
		expect(screen.getByText(`You're currently offline`)).toBeTruthy()
	})

	it(`should not display the toast if the user is online`, () => {
		jest.useFakeTimers()
		mockConnectionStatus(true)
		render(<NetInfoToast />)
		act(() => jest.runAllTimers())
		expect(screen.queryByTestId('toast-netinfo')).toBeNull()
	})

	it('should toggle visibility when connection status changes', () => {
		jest.useFakeTimers()
		// user is online
		mockConnectionStatus(true)
		const { rerender } = render(<NetInfoToast />)
		expect(screen.queryByText("You're currently offline")).toBeNull()

		// offline mode start
		mockConnectionStatus(false)
		rerender(<NetInfoToast />)
		act(() => jest.runAllTimers())
		expect(screen.getByTestId('toast-netinfo')).toBeTruthy()

		// offline mode end
		mockConnectionStatus(true)
		rerender(<NetInfoToast />)
		act(() => jest.runAllTimers())
		expect(screen.queryByTestId('toast-netinfo')).toBeNull()
	})

	it('should hide the toast after dismissing"', () => {
		jest.useFakeTimers()
		mockConnectionStatus(false)
		render(<NetInfoToast />)
		act(() => jest.runAllTimers())
		const toast = screen.getByTestId('toast-netinfo')
		expect(toast).toBeTruthy()
		fireEvent(toast, 'onDismiss')
		act(() => jest.runAllTimers())
		expect(screen.queryByTestId('toast-netinfo')).toBeNull()
	})
})
