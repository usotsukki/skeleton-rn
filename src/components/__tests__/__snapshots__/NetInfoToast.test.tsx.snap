// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetInfoToast should display the toast if the user is offline 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    {
      "elevation": undefined,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "transform": [
        {
          "translateY": 0,
        },
      ],
      "zIndex": 5,
    }
  }
  testID="toast-netinfo"
>
  <RCTSafeAreaView
    collapsable={false}
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        [
          {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        onLayout={[Function]}
        style={
          [
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        <View
          collapsable={false}
          jestAnimatedStyle={
            {
              "value": {
                "transform": [
                  {
                    "translateX": 0,
                  },
                  {
                    "translateY": 0,
                  },
                ],
              },
            }
          }
          style={
            {
              "transform": [
                {
                  "translateX": 0,
                },
                {
                  "translateY": 0,
                },
              ],
            }
          }
        >
          <View
            modifiers={
              {
                "alignments": {},
                "backgroundColor": undefined,
                "borderRadius": undefined,
                "color": undefined,
                "flexStyle": undefined,
                "margins": {},
                "paddings": {},
                "positionStyle": undefined,
                "typography": undefined,
              }
            }
            style={
              [
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
              ]
            }
          >
            <View
              style={
                [
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  [
                    {
                      "alignItems": "center",
                      "alignSelf": "center",
                      "backgroundColor": "#FFFFFF",
                      "borderRadius": 12,
                      "elevation": 3,
                      "flexDirection": "row",
                      "marginHorizontal": 20,
                      "marginVertical": 12,
                      "minHeight": 48,
                      "paddingLeft": 12,
                      "shadowColor": "#6E7881",
                      "shadowOffset": {
                        "height": 2,
                        "width": 0,
                      },
                      "shadowOpacity": 0.2,
                      "shadowRadius": 10,
                    },
                    undefined,
                    {
                      "backgroundColor": "rgba(0,0,0,0.9)",
                    },
                  ],
                ]
              }
            >
              <Image
                accessibilityRole="image"
                accessible={false}
                resizeMode="contain"
                source={1}
                style={
                  [
                    undefined,
                    undefined,
                    undefined,
                    false,
                    {
                      "height": 24,
                      "marginRight": 4,
                      "width": 24,
                    },
                  ]
                }
              />
              <View
                accessible={true}
                style={
                  [
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    {},
                    {},
                    {},
                    {
                      "flex": 1,
                      "justifyContent": "center",
                      "paddingVertical": 12,
                    },
                  ]
                }
              >
                <Text
                  accessibilityLabel="undefined notification, You're currently offline"
                  style={
                    [
                      {
                        "backgroundColor": "transparent",
                        "color": "#20303C",
                        "writingDirection": "ltr",
                      },
                      undefined,
                      undefined,
                      undefined,
                      undefined,
                      undefined,
                      undefined,
                      undefined,
                      undefined,
                      [
                        {
                          "color": "#20303C",
                          "marginLeft": 8,
                          "marginRight": 20,
                        },
                        {
                          "textAlign": "center",
                        },
                        {
                          "color": "white",
                        },
                      ],
                    ]
                  }
                  testID="toast-netinfo-message"
                >
                  You're currently offline
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;
